File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/api/ApiServer.java
package systems.zlink.samples.tictactoe.server.api;

import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.tictactoe.server.configuration.SampleLogging;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.configuration.ApiSettings;

public final class ApiServer {
    private ApiServer() {
    }

    public static ZLinkFrameworkConfigurer configure(ApiSettings settings) {
        return options -> {
            SampleLogging.configure(settings, "api");
            options.configureDispatch()
                .messageFlow(ZLinkMessageFlowLogMode.KEY_TRANSITIONS)
                .traceLogFile(SampleLogging.flowLogPath(settings, "api-" + settings.apiHttpPort()))
                .traceLabel("api-" + settings.apiHttpPort());
            options.configureLocations();
            options.addHandlersFromPackageOf(ApiServer.class);
            java.net.URI apiEndpoint = java.net.URI.create(settings.apiChannelEndpoint());
            options.addClientServerChannel(SampleNames.ApiChannel)
                .server()
                .setBindHost(apiEndpoint.getHost())
                .listen(apiEndpoint.getPort())
                .addHandlerGroup("api");
            ZLinkMeshNodeBuilder mesh = options.addRouteMesh(SampleNames.SpotMesh);
            mesh.listen(settings.routeEndpoint())
                .setRoutingIdPrefix("tictactoe-api");
            mesh.objects().client();
            for (String endpoint : settings.spotEndpoints()) {
                mesh.peerConnections().connect(endpoint);
            }
        };
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/api/ApiServer.kt
package systems.zlink.samples.kotlin.tictactoe.server.api

import kotlinx.coroutines.Dispatchers
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.configureDispatch
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleLogging
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.api.handlers.AuthenticatePlayerHandler
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerRes

object ApiServer {
    fun configure(settings: SampleSettings): ZLinkFrameworkConfigurer =
        ZLinkFrameworkConfigurer { options ->
            SampleLogging.configure(settings, "api")
            options.useCoroutineHandlers(Dispatchers.Default)
            options.addHandlersFromPackageOf(ApiServer::class.java)
            options.configureDispatch {
                messageFlow(ZLinkMessageFlowLogMode.KEY_TRANSITIONS)
                traceLogFile(SampleLogging.flowLogPath(settings, "api-${settings.apiHttpPort}"))
                traceLabel("api-${settings.apiHttpPort}")
            }
            options.configureLocations()
            val apiEndpoint = java.net.URI.create(settings.apiChannelEndpoint)
            options.addClientServerChannel(SampleNames.ApiChannel)
                .server()
                .setBindHost(apiEndpoint.host)
                .listen(apiEndpoint.port)
                .addHandlerGroup("api")
            val mesh = options.addRouteMesh(SampleNames.SpotMesh)
            mesh.listen(settings.routeEndpoint)
                .setRoutingIdPrefix("tictactoe-api")
            mesh.objects().client()
            settings.spotEndpoints.forEach { endpoint ->
                mesh.peerConnections().connect(endpoint)
            }
        }
}
